src/amm.rs

```diff
@@ -45,36 +46,60 @@ macro_rules! to_u128 {
     };
 }
 
+#[derive(PartialEq)]
+#[derive(Debug)]
+pub enum MathError {
+    DenominatorIsZero,
+    ResultOverflow
+}
+
+
 /// Calculating spot price given reserve of selling asset and reserve of buying asset.
 /// Formula : BUY_RESERVE * AMOUNT / SELL_RESERVE
 ///
 /// - `sell_reserve` - reserve amount of selling asset
 /// - `buy_reserve` - reserve amount of buying asset
 /// - `amount` - amount
 ///
-/// Returns None in case of error
-pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Option<Balance> {
-    ensure!(sell_reserve != 0);
+/// Returns MathError in case of error
+pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Result<Balance, MathError> {
+    if sell_reserve == 0 {
+        return Err(DenominatorIsZero);
+    }
+
+    if amount == 0 || buy_reserve == 0 {
```

Comment:
I would probably return error here as well. It was treated like an error.

Reply:
We can do that, however
one could argue that if Numerator is zero it might not be an exceptional state as we can still compute 0 / x. But if Denominator is zero we cannot compute x / 0. Maybe we should distinguish between those states.

Also I would like to point out that #calculate_sell_price will return Some if Numerator is zero. This might be inconsistent.
I will push a small test to show.

Reply:
Yes, now that i am thinking about it and see the usage , it is ok.